main/src/mill/modules/Jvm.scala

```diff
@@ -257,20 +261,52 @@ object Jvm {
     manifest.write(manifestOut)
     manifestOut.close()
 
-    def isSignatureFile(mapping: String): Boolean =
-      Set("sf", "rsa", "dsa").exists(ext => mapping.toLowerCase.endsWith(s".$ext"))
-
-    for(v <- classpathIterator(inputPaths)){
-      val (file, mapping) = v
-      val p = zipFs.getPath(mapping)
-      if (p.getParent != null) Files.createDirectories(p.getParent)
-      if (!isSignatureFile(mapping)) {
-        val outputStream = newOutputStream(p)
-        IO.stream(file, outputStream)
-        outputStream.close()
+    val concatEntries: mutable.Map[String, List[AssemblyEntry]] = mutable.Map.empty
+    val seen: mutable.Set[String] = mutable.Set.empty
+
+    val rulesMap = assemblyRules.collect {
+      case r@Rule.Append(path) => path -> r
+      case r@Rule.Exclude(path) => path -> r
+    }.toMap
+    val appendPatterns = assemblyRules.collect {
+      case Rule.AppendPattern(pattern) => pattern.asPredicate().test(_)
+    }
+    val excludePatterns = assemblyRules.collect {
+      case Rule.ExcludePattern(pattern) => pattern.asPredicate().test(_)
+    }
+
+    for(entry <- classpathIterator(inputPaths)) {
+      val mapping = entry.mapping
+
+      def append() = concatEntries += mapping -> (entry :: concatEntries.getOrElse(mapping, Nil))
+
+      rulesMap.get(mapping) match {
+        case Some(_:Assembly.Rule.Append) =>
+          append()
+        case Some(_:Assembly.Rule.Exclude) =>
+          ctx.log.info(s"Excluding entry: ${mapping}")
```

Comment:
I think we can skip the logging here; I find the logging from SBT's assembly task more annoying than useful